### DISCUSSION TITLE
Components diagram for the v6

### DISCUSSION THREAD

**Discussion by arturcic** [GitTools/GitVersion#3407] [Ideas]
Here is a conceptual components diagram that I think we should split the current code

![Class Diagram](https://www.plantuml.com/plantuml/png/VLB1hjCm4BpxAvRtplJ04LLLIv6uK77W2znauwomlPQzLeKG_qwSkAfJejvs9pFZcTZ793fATbPza14TGh2Ql4g_o4FaX4e9YKKuimliqKk2hwHDr4wfxewLhuuzw0JTCkttFXkKmm7UVwErV6l8RsUE-BQG8_vFXJ-W01fpRRjijL0zOLhX_JQIu9h6VY2JuuCu0tVgtpRlZomXIo7nChtA-oMJxK-cT2yAlKml6tuZ0NB1eflMET3Srv3KECyTlzDrqdpwEUeO2jFIrT1Mh5PqqLuRx2TXaRXbsvIjsvz2c8k4C35x5ey2mcmtkrpIt7y-8LtBk66lF0k2S00UwXFuN2v4DFa4v0NZe3jCC704bX7ZeqPIQdu77tFOxGvLhUfVNeFrijQWeMT4JSqUKCHsNtnRP35jaSbHgLgszMcmngj1MxyMhPv7z7rszXy0)

Those that are in yellow will be packaged as nuget packages.

Details:
- `GitVersion.Common` - will contain the interfaces needed for version calculation, for build agents and output, git interfaces as well as common infrastructure interfaces like ILogger, IFileSystem and so on.
- `GitVersion.Core` - will contain the implementation of infrastructure interfaces as well as the version calculation logic and configuration reading, writing, computing the effective configuration.
- `GitVersion.BuildAgents` - will contain different build agents implementations. The package will be referenced by `GitVersion.App` and `GitVersion.MsBuild`
- `GitVersion.Output` - will contain different output options implementations like file, envvars, a formatted string and so on. The package will be referenced by `GitVersion.App` and `GitVersion.MsBuild`
- `GitVersion.LibGit2Sharp` - will contain a libgit2sharp implementation of the git interfaces. The package will be referenced by `GitVersion.App` and `GitVersion.MsBuild`
- `GitVersion.ManagedGit` - will contain a managed git implementation of the git interfaces. The package will be referenced by `GitVersion.App` and `GitVersion.MsBuild`
- `GitVersion.App` - the dotnet tool that references `GitVersion.Core`  as libs and `GitVersion.BuildAgents`, `GitVersion.Output` and (`GitVersion.LibGit2Sharp` _or_ `GitVersion.ManagedGit`)
- `GitVersion.MsBuild` - similar to `GitVersion.App`

In case someone wants to use the GitVersion as a lib and use the version calculation in their code, they will be able to reference the 
`GitVersion.Core`, either `GitVersion.LibGit2Sharp` _or_ `GitVersion.ManagedGit` and optionally `GitVersion.BuildAgents` and `GitVersion.Output`. The consumer will be able to provide he's own BuildAgent implementation or output format they need

**Comment by HHobeck** [comment]
I'm not sure if common is a good naming. Maybe Contract or Abstractions? Everything else looks good for me.

**Comment by arturcic** [comment]
Naming is hard, you know ;), but I agree,

**Comment by arturcic** [comment]
My question is what should we do with `GitVersion.Configuration` - should it be a lib dependency for `GitVersion.Core`? In that case `GitVersion.Configuration` will implement the Configuration interfaces defined in `GitVersion.Abstractions` and then in `GitVersion.Core` we consume only those Configuration interfaces that will provide the EffectiveConfiguration? Is it EffectiveConfiguraton enough or we still need GitVersionConfiguration as well?

**Reply by HHobeck** [reply]
Very good question. For me it is fine if we do this like you described it. In my opinion we need the GitVersionConfiguration as well. You can think of the GitVersionConfiguration+BranchConfigration=EffectiveConfiguraton and Branch+EffectiveConfiguraton=EffectiveBranchConfiguraton. If we are iterating to the branches then we have sometimes more then one EffectiveBranchConfigurations to determine the next semantic version. Thus it needs to be present in the Core library.

**Reply by HHobeck** [reply]
EffectiveBranchConfiguraton is maybe not named correct or even temporary and can be refactored/removed. But at this moment we need the branch information

**Reply by arturcic** [reply]
> GitVersionConfiguration+BranchConfigration=EffectiveConfiguraton and Branch+EffectiveConfiguraton=EffectiveBranchConfiguraton. 

Can we include the branch information directly in the EffectiveConfiguraton ?

**Reply by HHobeck** [reply]
> Can we include the branch information directly in the EffectiveConfiguraton ?

Of course we can do this. :)

**Comment by arturcic** [comment]
Updated
![Class Diagram](https://www.plantuml.com/plantuml/png/VLBDojim4BphARR9kJdqM498ca8l3Jtq2MHxROjAMY6j2ANqtJ_zO2n_dmdui3kQcTsnT0qiFKUZnIUUqI0uBPKLuO-ou2cW4AnO8zp9EB9eES1t9ISlZH3zWi6j2-nbpufi09c-jZ-VRPoGBnSu_aMjwTa83tVoUCYIFXKVgV0F1C3AlqlTHospSH6b5exFMJ5kQMH7DKM_4WkmKF-tk_-Ax29d4jNgrRxVejB3RKfbpLchbmb_A0Pbd4QJsh9GoJibkAxpFkDFrINDbz-pz2upjUeIUMT8Atf8AoSSaZ1BJEtsPPlqDrTcIESISo2soC14UZVB8vJyPKBeShbZBomn0fC36ZSFuMk-5fRA1b2MqO-onm0ZUI2UqQzXWX3bDNoExdIwR3n4_UExH_NwTeycjO-Bb7ivo2FRlavgaJgWHPAl4CklQ7mQhF5gqCQlGLVFAzeX6lq6)

[Edit diagram](https://www.plantuml.com/plantuml/uml/VP7FJiCm3CRlUGfhdDc18vgc3P2uC77W2RBLpIpo9qeSJGZnxZHDK9jjx1R_-dt-x7eJMGPEHeluIHQ2YoW44sk4PsUyisWvmWj95QGHul37ybVZBCW8X_9QhMnIoEirt7sXrkxKY0D2qGOSb9hsMNSw4cCZipsf52HJRJs2_mtlYNtYh7JbTMk8fqIwswgyK7R8yhhbMBmI0nclqGpbEDKYMnLndGM-GG1CZZVQ3vw7ZwCCFYk7kRFu8gGnxQIL2hl1c2scLDTjF_CJRFsO8RqVb9rZYmpidBwwoowE-uy9SJy-BziBwnY1dGVNrxi-vgkm91k1B6FevG4Zz2w0uoE6QOqen7ZT--INotLrY_gNMrYFrK95vwHUluL2UZ-XWFEwXCr9IPeJHKrPCeIe_wJftBBQVQ9DmXbLDA4sQOEsIqR_0W00)

**Reply by asbjornu** [reply]
I think `Abstractions` belong inside the `Core` and don't really need its separate module or namespace. Abstractions are visibly so by them either being `abstract class` or `interface`, no need to underscore this any further, imho.

**Reply by asbjornu** [reply]
I think we need to detail what goes inside each box to reach an agreement whether the boxes should be one way or another. I think the interfaces of everything required to do version calculation alongside the version calculation itself belongs in `Core` and I think it should be named `Core` so it's obvious when trying to visualise the project where the project is in relation to the other projects.

**Reply by arturcic** [reply]
@asbjornu can you [edit this diagram](https://www.plantuml.com/plantuml/uml/VP7FJiCm3CRlUGfhdDc18vgc3P2uC77W2RBLpIpo9qeSJGZnxZHDK9jjx1R_-dt-x7eJMGPEHeluIHQ2YoW44sk4PsUyisWvmWj95QGHul37ybVZBCW8X_9QhMnIoEirt7sXrkxKY0D2qGOSb9hsMNSw4cCZipsf52HJRJs2_mtlYNtYh7JbTMk8fqIwswgyK7R8yhhbMBmI0nclqGpbEDKYMnLndGM-GG1CZZVQ3vw7ZwCCFYk7kRFu8gGnxQIL2hl1c2scLDTjF_CJRFsO8RqVb9rZYmpidBwwoowE-uy9SJy-BziBwnY1dGVNrxi-vgkm91k1B6FevG4Zz2w0uoE6QOqen7ZT--INotLrY_gNMrYFrK95vwHUluL2UZ-XWFEwXCr9IPeJHKrPCeIe_wJftBBQVQ9DmXbLDA4sQOEsIqR_0W00) and provide your visions?

**Reply by HHobeck** [reply]
Okay never mind. So you mean we are skipping the hole Abstractions.dll or Common.dll idea and put everything in one Core.dll? As a consequence I want to point out that all other projects have in this design a reference to the Core.dll. Is this assumption correct?

**Reply by HHobeck** [reply]
It might be possible as long we have no circular references.

**Reply by asbjornu** [reply]
Yes, I prefer `Core` as it explains the project's purpose in the architecture much more explicitly and precisely than `Common`. `Core` should not reference anything whatsoever other than `mscorlib.dll`. All other projects in the solution should reference `Core`, but not the other way around.

**Reply by asbjornu** [reply]
> @asbjornu can you [edit this diagram](https://www.plantuml.com/plantuml/uml/VP7FJiCm3CRlUGfhdDc18vgc3P2uC77W2RBLpIpo9qeSJGZnxZHDK9jjx1R_-dt-x7eJMGPEHeluIHQ2YoW44sk4PsUyisWvmWj95QGHul37ybVZBCW8X_9QhMnIoEirt7sXrkxKY0D2qGOSb9hsMNSw4cCZipsf52HJRJs2_mtlYNtYh7JbTMk8fqIwswgyK7R8yhhbMBmI0nclqGpbEDKYMnLndGM-GG1CZZVQ3vw7ZwCCFYk7kRFu8gGnxQIL2hl1c2scLDTjF_CJRFsO8RqVb9rZYmpidBwwoowE-uy9SJy-BziBwnY1dGVNrxi-vgkm91k1B6FevG4Zz2w0uoE6QOqen7ZT--INotLrY_gNMrYFrK95vwHUluL2UZ-XWFEwXCr9IPeJHKrPCeIe_wJftBBQVQ9DmXbLDA4sQOEsIqR_0W00) and provide your visions?

Yes, will do.

**Comment by arturcic** [comment]
Well the current state of the code base: we have `Core` that has basically every interfaces, then we have the `Output`, `LibGit2Sharp` and `BuildAgents` modules that depend on `Core` where their interfaces are defined, and implement in each specific module. 

Then we have the consumer apps or the packages/tools that the GitVersion user will actually use - `GitVersion.Tool` and `MsBuild`.

These 2 apps will reference the `Core` and the modules.

The idea behind the `Abstraction` was to move the interfaces to the new module out of `Core` so the the `Output`, `LibGit2Sharp` and `BuildAgents` do not rely on the `Core` for the interfaces. The problem I see in the current state is the `Core` beside the interfaces for the modules (aka contracts) also contain the version calculation logic. So we need to make a design that does not couple the modules to the `Core` (that is kind of Contracts + Version calculation). @HHobeck @asbjornu let me know what are your thoughts on this?

**Comment by asbjornu** [comment]
Okay, I've finally found some time to take a look at this. I don't think we need to adjust the proposed diagram much. 

I suggest we do is merge `Common` into `Core`. `Core` should not have any external dependencies and I don't think we need another common project in addition to `Core`.

I'm also a bit unsure about whether we want to publish all of the projects marked with `nuget` as NuGet packages. Isn't the dotnet tool, standalone CLI and MSBuild Task enough?

Lastly, `GitVersion.Configuration` seems to be missing from your explanation. I've included it in the table below.

![GitVersion components diagram](https://www.plantuml.com/plantuml/svg/TP71Jlim3CRlUGfh__D_1uwecZP0uWBYm1EahPjPf76KE9eGujrfcgAb5RlPNt-VlyO-Hj51qcXL_207WICg9IGMuP57pmwTH7WYRO8UbUf-jUbhG00TeSj5qrXg3mVuTpcJu0fp0vaKj12xWcVXjk4jYK-IIIxLJV8XaUrF9lzWndMfRkAxPnAWqLiSftR-drts6P85XYzG05V72xMJv-xzh8FFv3Gumvk4sl6gdJROJwxCZwNxm_DTR_Bav-dQ-mdhMHmA2BFD-2TQonTL9SJvsVFusCvbqxXaK2fSEHO4OGyypFUvptSGJMuvLviYEOmHB1lgGBiUo0c6GNSOOU00B6SCrvT5fPRz_az-lp_COrIvq4eojExBkLUIoY-jXQel8RLICch5qAej2KejwwWdRxKoVQkMXArgg9Rgf2EwFets1m00)


| Project                    | Description |
| -------------------------: | :---------- |
| `GitVersion.Core`          | Contains the interfaces needed for version calculation, for build agents and output, git interfaces as well as common infrastructure interfaces like ILogger, IFileSystem and so on. Also contains all objects and business logic for version calculation. |
| `GitVersion.Configuration` | Contains the implementation of the configuration interfaces, with external dependency on YamlDotNet. |
| `GitVersion.BuildAgents`   | Contains different build agents implementations. The package will be referenced by `GitVersion.App` and `GitVersion.MsBuild`.
| `GitVersion.Output`        | Contains different output options implementations like file, envvars, a formatted string and so on. Will have a dependency on Newtonsoft.Json (probably). Will be referenced by `GitVersion.App` and `GitVersion.MsBuild`.
| `GitVersion.LibGit2Sharp`  | Contains a reference to LibGit2Sharp, implementing the Git interfaces. The package will be referenced by `GitVersion.App` and `GitVersion.MsBuild`.
| `GitVersion.ManagedGit`    | Contains a managed implementation of the Git interfaces. Might be merged into `Core` since it shouldn't require any external dependencies. The package will be referenced by `GitVersion.App` and `GitVersion.MsBuild`.
| `GitVersion.App`           | The dotnet tool that references `GitVersion.Core`  as libs and `GitVersion.BuildAgents`, `GitVersion.Output` and (`GitVersion.LibGit2Sharp` _or_ `GitVersion.ManagedGit`). |
| `GitVersion.MsBuild`       | similar to `GitVersion.App`. |
